### PR TITLE
Use stree for message block `fss`

### PR DESCRIPTION
**Pull request by neilalexander**
Should provide some deduplication of long subjects in memory when message blocks are loaded in.

Signed-off-by: Neil Twigg <neil@nats.io>

### CODE REVIEW

File: server/filestore.go [hunk 1]
```diff
@@ -2345,26 +2347,27 @@ func (mb *msgBlock) firstMatching(filter string, wc bool, start uint64, sm *Stor
...
 	}
 
 	if !doLinearScan {
 		// If we have a wildcard match against all tracked subjects we know about.
 		if wc {
 			subs = subs[:0]
-			for subj := range mb.fss {
-				if isMatch(subj) {
-					subs = append(subs, subj)
+			mb.fss.Iter(func(bsubj []byte, _ *SimpleState) bool {
```
Comment (derekcollison) [inline server/filestore.go:2357]: Just do a match here??

Comment (derekcollison) [reply server/filestore.go:2357]:
```go
mb.fss.Match(stringToBytes(filter))
```

File: server/filestore.go [hunk 2]
```diff
@@ -2485,8 +2488,8 @@ func (mb *msgBlock) filteredPendingLocked(filter string, wc bool, sseq uint64) (
 	}
 
 	var havePartial bool
-	for subj, ss := range mb.fss {
-		if isAll || isMatch(subj) {
+	mb.fss.Iter(func(bsubj []byte, ss *SimpleState) bool {
```
Comment (derekcollison) [inline server/filestore.go:2491]: Do Match..

File: server/filestore.go [hunk 3]
```diff
@@ -2711,7 +2715,8 @@ func (fs *fileStore) SubjectsState(subject string) map[string]SimpleState {
 		}
 		// Mark fss activity.
 		mb.lsts = time.Now().UnixNano()
-		for subj, ss := range mb.fss {
+		mb.fss.Iter(func(bsubj []byte, ss *SimpleState) bool {
```
Comment (derekcollison) [inline server/filestore.go:2718]: Same.

File: server/filestore.go [hunk 4]
```diff
@@ -3008,7 +3014,8 @@ func (fs *fileStore) NumPending(sseq uint64, filter string, lastPerSubject bool)
 			mb.lsts = time.Now().UnixNano()
 
 			var havePartial bool
-			for subj, ss := range mb.fss {
+			mb.fss.Iter(func(bsubj []byte, ss *SimpleState) bool {
```
Comment (derekcollison) [inline server/filestore.go:3017]: Match.

**Review by derekcollison — CHANGES_REQUESTED** [review]

**Review by wallyqs — APPROVED** [review]
LGTM

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**